File: backend_old/services/complianceService.js
// CIS-style compliance checks against the Neo4j graph

const { runQuery } = require('../db/neo4j');

const CIS_CHECKS = [
  {
    id: 'CIS-1',
    title: 'No public access to compute resources',
    description: 'Compute instances should not be directly reachable from the public internet without firewall controls.',
    query: `
      MATCH (internet:Resource {id: 'internet'})
      -[:EXPOSED_TO|CONNECTS_TO*1..2]->(c:Resource {type: 'compute'})
      RETURN c.id AS id, c.name AS name, c.cloud AS cloud
    `,
    failMessage: (row) => `${row.cloud?.toUpperCase()} compute resource '${row.name}' is publicly reachable from the Internet`,
  },
  {
    id: 'CIS-2',
    title: 'IAM least privilege — no wildcard/admin roles on public compute',
    description: 'Privileged IAM roles should not be attached to internet-exposed compute instances.',
    query: `
      MATCH (internet:Resource {id: 'internet'})
      -[:EXPOSED_TO|CONNECTS_TO*1..2]->(c:Resource {type: 'compute'})
      -[:HAS_ROLE]->(r:Resource {type: 'identity'})
      WHERE r.privilegedRole = true
      RETURN r.id AS id, r.name AS name, r.cloud AS cloud, c.name AS computeName
    `,
    failMessage: (row) => `Privileged role '${row.name}' is attached to internet-reachable compute '${row.computeName}'`,
  },
  {
    id: 'CIS-3',
    title: 'Sensitive data not accessible from privileged public compute',
    description: 'Sensitive data stores should not be accessible via roles attached to internet-exposed compute.',
    query: `
      MATCH (internet:Resource {id: 'internet'})
      -[:EXPOSED_TO|CONNECTS_TO*1..2]->(c:Resource {type: 'compute'})
      -[:HAS_ROLE]->(r:Resource {type: 'identity'})
      -[:CAN_ACCESS]->(d:Resource {sensitive: true})
      RETURN d.id AS id, d.name AS name, d.cloud AS cloud
    `,
    failMessage: (row) => `Sensitive data store '${row.name}' is transitively accessible from the Internet`,
  },
  {
    id: 'CIS-4',
    title: 'No open firewall rules (0.0.0.0/0)',
    description: 'Security groups and firewall rules should not allow unrestricted inbound traffic.',
    query: `
      MATCH (n:Resource)
      WHERE n.openFirewall = true
      RETURN n.id AS id, n.name AS name, n.cloud AS cloud
    `,
    failMessage: (row) => `${row.cloud?.toUpperCase()} firewall/security group '${row.name}' allows unrestricted 0.0.0.0/0 ingress`,
  },
];

async function runAllChecks() {
  const results = [];
  for (const check of CIS_CHECKS) {
    try {
      const records = await runQuery(check.query);
      const violations = records.map((r) => {
        const row = {};
        r.keys.forEach((k) => { row[k] = r.get(k); });
        return row;
      });

      results.push({
        id: check.id,
        title: check.title,
        description: check.description,
        status: violations.length === 0 ? 'pass' : 'fail',
        violations: violations.map((v) => ({
          resourceId: v.id,
          message: check.failMessage(v),
        })),
      });
    } catch (err) {
      results.push({
        id: check.id,
        title: check.title,
        description: check.description,
        status: 'error',
        violations: [],
        error: err.message,
      });
    }
  }
  return results;
}

function mapFindingToCompliance(findingType) {
  const mapping = {
    'public-compute': ['CIS-1', 'CIS-4'],
    'privileged-role': ['CIS-2'],
    'sensitive-data-exposure': ['CIS-3'],
  };
  return mapping[findingType] || [];
}

module.exports = { runAllChecks, mapFindingToCompliance, CIS_CHECKS };

File: backend/services/compliance_service.go
package services

import "hawkeye-backend/models"

// EvaluateCompliance evaluates an asset against SOC2 / CIS benchmark rules
// and returns the full list of results (PASS and FAIL).
func EvaluateCompliance(asset *models.Asset) []models.ComplianceResult {
	var results []models.ComplianceResult

	// CIS-1 – No public access to compute resources
	cis1 := models.ComplianceResult{
		ID:          "CIS-1",
		Title:       "No public access to compute resources",
		Status:      "PASS",
		Description: "Compute resources must not be directly accessible from the internet.",
	}
	if asset.IsPublic {
		cis1.Status = "FAIL"
		cis1.Description = "Asset is publicly accessible. Remove public IP or add firewall rules."
	}
	results = append(results, cis1)

	// CIS-2 – IAM role must follow least privilege
	cis2 := models.ComplianceResult{
		ID:          "CIS-2",
		Title:       "IAM role violates least privilege",
		Status:      "PASS",
		Description: "IAM roles must follow the principle of least privilege.",
	}
	if asset.HasAdminRole {
		cis2.Status = "FAIL"
		cis2.Description = "Asset has an admin IAM role attached. Downscope permissions to required actions only."
	}
	results = append(results, cis2)

	// CIS-3 – Sensitive data must not be exposed publicly
	cis3 := models.ComplianceResult{
		ID:          "CIS-3",
		Title:       "Sensitive data exposed publicly",
		Status:      "PASS",
		Description: "Assets containing sensitive data must not be publicly reachable.",
	}
	if asset.IsSensitive && asset.IsPublic {
		cis3.Status = "FAIL"
		cis3.Description = "Asset holds sensitive data and is publicly accessible. Restrict network access immediately."
	}
	results = append(results, cis3)

	// CIS-4 – Open firewall (0.0.0.0/0)
	cis4 := models.ComplianceResult{
		ID:          "CIS-4",
		Title:       "Open firewall (0.0.0.0/0)",
		Status:      "PASS",
		Description: "Firewall rules must not allow unrestricted inbound traffic.",
	}
	if asset.IsPublic {
		// Treat any publicly reachable resource as having an overly permissive firewall
		cis4.Status = "FAIL"
		cis4.Description = "Asset is reachable from 0.0.0.0/0. Restrict inbound rules to known CIDRs."
	}
	results = append(results, cis4)

	return results
}
